### PR TITLE
Add new diagnostic writer using annotate-snippet library

### PR DESCRIPTION
**Pull request by phansch**
This adds a new diagnostic writer `AnnotateRsEmitterWriter` that uses
the [`annotate-snippet`][as] library to print out the human readable
diagnostics.

The goal of #59346 is to eventually switch over to using the library instead of
maintaining our own diagnostics output.

This PR does **not** add all the required features to the new
diagnostics writer. It is only meant as a starting point so that other
people can start contributing as well.

There are some FIXMEs in `librustc_errors/annotate_rs_emitter.rs` that
point at yet to be implemented features of the new diagnostic emitter, however
those are most likely not exhaustive.

[as]: https://github.com/rust-lang/annotate-snippets-rs

### CODE REVIEW

File: src/test/ui/annotate-snippet/missing-type.rs [hunk 1]
```diff
@@ -0,0 +1,5 @@
+// compile-flags: --error-format human-annotate-rs
+
+pub fn main() {
+    let x: Iter;
+}
```
Comment (estebank) [inline src/test/ui/annotate-snippet/missing-type.rs]: You need to run `./x.py test src/test/ui --stage 1 --bless` for the tests to pass.

File: src/test/ui/annotate-snippet/missing-type.rs [hunk 2]
```diff
@@ -0,0 +1,5 @@
+// compile-flags: --error-format human-annotate-rs
+
+pub fn main() {
+    let x: Iter;
```
Comment (estebank) [inline src/test/ui/annotate-snippet/missing-type.rs]:
```suggestion
    let x: Iter; //~ ERROR cannot find type `Iter` in this scope
```

File: src/librustc_errors/annotate_rs_emitter.rs [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    ) -> SourceAnnotation {
+        SourceAnnotation {
+            range: (annotation.start_col, annotation.end_col),
+            label: annotation.label.unwrap_or("".to_string()),
+            annotation_type: Self::annotation_type_for_level(self.level)
+        }
+    }
+
+    /// Provides the source string for the given `line` of `file`
+    fn source_string(file: Lrc<SourceFile>,
```
Comment (oli-obk) [inline src/librustc_errors/annotate_rs_emitter.rs]: use block-indent instead of visual indent

File: src/librustc_errors/annotate_rs_emitter.rs [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+                     line: &Line) -> String {
+        let source_string = match file.get_line(line.line_index - 1) {
```
Comment (oli-obk) [inline src/librustc_errors/annotate_rs_emitter.rs]: I think this function's entire body can be `file.get_line(line.line_index - 1).map(|s| s.to_string()).unwrap_or_default()`, but I'm not sure what `get_line` returns.

File: src/librustc_errors/annotate_rs_emitter.rs [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+    // FIXME(#59346): This method is used via the public interface, but setting the `ui_testing`
+    // flag currently does not anonymize line numbers. We would have to add the `maybe_anonymized`
+    // method from `emitter.rs` and implement rust-lang/annotate-snippets-rs#2 in order to
+    // anonymize line numbers.
+    pub fn ui_testing(mut self, ui_testing: bool) -> Self {
+        self.ui_testing = ui_testing;
+        self
+    }
+
+    fn emit_messages_default(&mut self,
```
Comment (oli-obk) [inline src/librustc_errors/annotate_rs_emitter.rs]: use block-indent

File: src/librustc_errors/annotate_rs_emitter.rs [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+            children,
+            suggestions
+        };
+        if let Some(snippet) = converter.to_annotation_snippet() {
+            let dl = DisplayList::from(snippet);
+            let dlf = DisplayListFormatter::new(true);
+            // FIXME(#59346): Figure out if we can _always_ print to stderr or not.
+            // `emitter.rs` has the `Destination` enum that lists various possible output
+            // destinations.
+            eprint!("{}", dlf.format(&dl));
```
Comment (oli-obk) [inline src/librustc_errors/annotate_rs_emitter.rs]: use `eprintln` to get rid of the missing trailing newline icon shown by github

File: src/tools/tidy/src/deps.rs [hunk 7]
```diff
@@ -62,6 +62,8 @@ const WHITELIST_CRATES: &[CrateVersion<'_>] = &[
 const WHITELIST: &[Crate<'_>] = &[
     Crate("adler32"),
     Crate("aho-corasick"),
+    Crate("annotate-snippets"),
+    Crate("ansi_term"),
```
Comment (oli-obk) [inline src/tools/tidy/src/deps.rs]: cc @rust-lang/compiler 	we're pulling in a new dependency here, and we have no control over it: `ansi_term` (`annotate-snippets` is also a new dep, but it is under the `rust-lang` org: https://github.com/rust-lang/annotate-snippets-rs).

**Review by oli-obk — CHANGES_REQUESTED** [review]
awesome, just some nits and maybe opinions on the `ansi_term` transitive dependency.

File: src/tools/tidy/src/deps.rs (hunk 7 above)
Comment (nagisa) [reply src/tools/tidy/src/deps.rs]: Two points in favour of allowing it: its license (MIT) is in our list of allowed licenses for external crates; and it is already a transitive dependency of our other core projects such as `cargo`.

As long as changes to the `ansi_term` are vetted as part of the `annotate-snippets` crate and the crate is locked to a specific patch version I’m in favour of the dependency.

Comment (phansch) [reply src/tools/tidy/src/deps.rs]: The usage of `ansi_term` in `annotate-snippets` seems pretty lightweight so far and we could probably switch to another one as well, if that's a concern.

**Comment by phansch** [comment]
Rebased.

**Pull request merged**